### PR TITLE
k8s: add K8s endpoints equalness and missing function

### PR DESCRIPTION
**Pull request by aanm**
Read per commit basis.

```release-note
Stop triggering unnecessary loadbalancer setups for every Kubernetes endpoint watch event.
```

---
This change is [Reviewable](https://reviewable.io/reviews/cilium/cilium/5642)

### CODE REVIEW

File: pkg/k8s/factory_functions_test.go [hunk 1]
```diff
@@ -168,3 +169,95 @@ func (s *K8sSuite) Test_equalV2CNP(c *C) {
...
+						Name: "rule1",
+					},
+					Subsets: []core_v1.EndpointSubset{
+						{
+							Addresses: []core_v1.EndpointAddress{
+								{
+									IP: "172.0.0.1",
+								},
+							},
+						},
```
Comment (tgraf) [inline pkg/k8s/factory_functions_test.go:252]: Please add a test with multiple IPs

File: daemon/k8s_watcher.go [hunk 2]
```diff
@@ -994,6 +992,44 @@ func (d *Daemon) deleteK8sEndpointV1(ep *v1.Endpoints) {
 	endpointMetadataCache.delete(ep)
 }
 
+func (d *Daemon) missingK8sEndpointsV1(m versioned.Map) versioned.Map {
```
Comment (tgraf) [inline daemon/k8s_watcher.go]: Please add a description of what dies function implements.

**Review by tgraf — CHANGES_REQUESTED** [review]
Awesome. This looks great and brings us a lot closer to moving more code into `pkg/k8s`.

**Comment by aanm** [comment]
>Awesome. This looks great and brings us a lot closer to moving more code into pkg/k8s.

I've noticed that, some of the functions depend on daemon only to access the loadbalancer. Once the loadbalancer will be in its package it will be a lot easier.

**Comment by aanm** [comment]
test-me-please

k8s tests timed out

File: pkg/loadbalancer/loadbalancer.go [hunk 3]
```diff
@@ -115,7 +115,7 @@ type LoadBalancer struct {
 	SVCMapID  SVCMapID
 	RevNATMap RevNATMap
 
-	K8sMU        lock.Mutex
+	K8sMU        lock.RWMutex
```
Comment (jrajahalme) [inline pkg/loadbalancer/loadbalancer.go:118]: This change seems unrelated, is `K8sMU` ever read locked?

Comment (jrajahalme) [reply pkg/loadbalancer/loadbalancer.go:118]: Never mind, saw the RLock in the last commit.

File: daemon/k8s_watcher.go [hunk 4]
```diff
@@ -982,6 +992,46 @@ func (d *Daemon) deleteK8sEndpointV1(ep *v1.Endpoints) {
...
+	for _, metaEP := range metaEPs {
+		lbEP, ok := d.loadBalancer.K8sEndpoints[metaEP.svcNS]
+		if !ok {
+			missing.Add(metaEP.uuid, metaEP.object)
+			continue
+		}
+		if !metaEP.k8sSvcEP.DeepEqual(lbEP) {
+			missing.Add(metaEP.uuid, metaEP.object)
+		}
+	}
```
Comment (jrajahalme) [inline daemon/k8s_watcher.go:1030]: The case where `d.loadBalancer.K8sEndpoints` has endpoints not in `m` is not interesting here?

File: daemon/k8s_watcher.go [hunk 5]
```diff
@@ -474,9 +474,7 @@ func (d *Daemon) EnableK8sWatcher(reSyncPeriod time.Duration) error {
 					return nil
 				}
 			},
-			func(m versioned.Map) versioned.Map {
-				return m
-			},
+			d.missingK8sEndpointsV1,
```
Comment (jrajahalme) [inline daemon/k8s_watcher.go:477]: I guess this is the crux of the matter here, instead of considering all of the endpoints in `m` missing, we now actually check which ones are missing.

File: daemon/k8s_watcher.go [hunk 6]
```diff
@@ -982,6 +992,46 @@ func (d *Daemon) deleteK8sEndpointV1(ep *v1.Endpoints) {
 	endpointMetadataCache.delete(ep)
 }
 
+// missingK8sEndpointsV1 returns a map containing missing endpoints considered
+// missing from the k8sEndpoints of the loadbalancer.
+func (d *Daemon) missingK8sEndpointsV1(m versioned.Map) versioned.Map {
```
Comment (jrajahalme) [inline daemon/k8s_watcher.go:997]: It would be nice if the name of this function indicated that it's working on the loadbalancer endpoints, e.g., `missingLoadBalancerK8sEndpointsV1()`

**Review by jrajahalme — APPROVED** [review]
Nothing blocking from me.

File: daemon/k8s_watcher.go (hunk 4 above)
Comment (aanm) [reply daemon/k8s_watcher.go:1030]: @jrajahalme No, because the user might add some endpoints manually and we don't want to overwrite them. There's another place that keeps everything in sync with kubernetes. https://github.com/cilium/cilium/blob/89cd59b80019770903fb1e87becddde8c33c0815/pkg/k8s/utils/factory.go#L237

File: daemon/k8s_watcher.go (hunk 5 above)
Comment (aanm) [reply daemon/k8s_watcher.go:477]: Correct.

File: daemon/k8s_watcher.go (hunk 6 above)
Comment (aanm) [reply daemon/k8s_watcher.go:997]: I can do it later on for all the `missing` functions.

**Review by tgraf — APPROVED** [review]

**Comment by tgraf** [comment]
test-me-please

k8s tests got EOF -> https://github.com/cilium/cilium/issues/5648

**Comment by aanm** [comment]
test-missed-k8s
timed-out

**Comment by aanm** [comment]
test-me-please

k8s tests got EOF -> https://github.com/cilium/cilium/issues/5648

**Comment by aanm** [comment]
test-missed-k8s
